packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -507,6 +528,21 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
     return Future<void>.value();
   }
 
+  final Map<PreHotRestartCallback, String?> _hotRestartCallbacks = <PreHotRestartCallback, String?>{};
+
+  /// Register a callback that will be invoked before a hot restart is called.
+  ///
+  /// In non-debug modes this method is a no-op.
+  ///
+  /// This can be used to release native resources acquired through platform
+  /// channels or `dart:ffi`.
+  void registerHotRestartCallback(PreHotRestartCallback callback, {String? debugLabel}) {
```

Comment:
I could give a much larger example of how to use this based on some dart:ffi native code (since I have an implementation that uses this sitting right here...)

Reply:
the more documentation the better!

Reply:
done-ish (could add the native code example too)